File: 12 Euclidean algorithm/ChocolatesByNumbers/CBN.js
function solution(N, M) { // Integer N represents the number of chocolates
  if (N === 1) return 1;
  if (M === 1) return N;
  const gcd = (n, m) => (m === 0 ? n : gcd(m, n % m));
  return N / gcd(N, M);
}
console.time('time');
solution(10, 4); // 5 eating 0, 4, 8, 2, 6
solution(415633212, 234332);
console.timeEnd('time');
// O(log(N + M)) 

File: 12 Euclidean algorithm/ChocolatesByNumbers/cbn.java
class Solution {
  public int solution(int N, int M) {
    return N / (gcd(N, M));
  }

  public static int gcd(int n, int m) {
    if (n % m == 0)
      return m;
    else
      return gcd(m, n % m);
  }
}
